### PR TITLE
Metrics helpers

### PR DESCRIPTION
**Pull request by kylebgorman**
This adds a small `metrics` struct which stores the information that `loss` should be minimized, `accuracy` should be maximized, and generates appropriate file names for checkpoints. It is then used in `train.py` to implement the checkpointing callback and the patience callback, and in `schedulers.py` to implement the reduce-on-plateau scheduler.

Whereas #171 used the same flag (`--val_metric`) to do checkpointing and patience, offline discussion concluded that we want these to be separate. Naturally the reduce-on-plateau scheduler was also left separate. 

Some flags have been renamed, and I am happy to discuss this further:

* `--val_metric` is now `--checkpoint_metric` and `--patience_metric`.
* `--reduceonplateau_mode` is now `--reduceonplateau_metric` to match the previous.
* I am still bothered by `--save_top_k` so I have proposed another alternative: `--num_checkpoints`. Hard to get clearer than that.

Since there has been a decent amount of plumbing changes here, I'd appreciate a closer review. However, my local tests seem just fine. 

This PR is a followup to PR #171 and a further response to issue #170.

### CODE REVIEW

File: yoyodyne/train.py [hunk 1]
```diff
@@ -36,59 +36,56 @@ def _get_logger(experiment: str, model_dir: str, log_wandb: bool) -> List:
...
     """Creates the callbacks.
 
     We will reach into the callback metrics list to picks ckp_callback to find
     the best checkpoint path.
 
     Args:
-        patience (int, optional).
+        num_checkpoints (int, optional): number of checkpoints to save. To
+            save one checkpoint per epoch, use `-1`.
+        checkpoint_metric (string, optional): the
```
Comment (Adamits) [inline yoyodyne/train.py:52]: Missing comment!

File: yoyodyne/train.py [hunk 2]
```diff
@@ -36,59 +36,56 @@ def _get_logger(experiment: str, model_dir: str, log_wandb: bool) -> List:
...
+        patience (int, optional): number of epochs with no
+            progress (according to `patience_metric`) before triggering
+            early stopping.
+
+
         val_metric (str, optional).
-        save_top_k (int, optional)....
+        save_top_k (int, optional).
```
Comment (Adamits) [inline yoyodyne/train.py:59]: delete

File: yoyodyne/train.py [hunk 3]
```diff
@@ -36,59 +36,56 @@ def _get_logger(experiment: str, model_dir: str, log_wandb: bool) -> List:
...
-        patience (int, optional).
+        num_checkpoints (int, optional): number of checkpoints to save. To
+            save one checkpoint per epoch, use `-1`.
+        checkpoint_metric (string, optional): the
+        patience (int, optional): number of epochs with no
+            progress (according to `patience_metric`) before triggering
+            early stopping.
+
+
         val_metric (str, optional).
```
Comment (Adamits) [inline yoyodyne/train.py:58]: delete

File: yoyodyne/train.py [hunk 4]
```diff
@@ -36,59 +36,56 @@ def _get_logger(experiment: str, model_dir: str, log_wandb: bool) -> List:
...
 
     We will reach into the callback metrics list to picks ckp_callback to find
     the best checkpoint path.
 
     Args:
-        patience (int, optional).
+        num_checkpoints (int, optional): number of checkpoints to save. To
+            save one checkpoint per epoch, use `-1`.
+        checkpoint_metric (string, optional): the
+        patience (int, optional): number of epochs with no
```
Comment (Adamits) [inline yoyodyne/train.py:53]: 1) I think we need a comment for `patience_metric`
2) Should patience be an attribute of the patience metric object? I think probably no given the design, but just want to double check this is intentional.

Comment (Adamits) [reply yoyodyne/train.py:53]: Actually ignore 2)---this is synonymous to how we have a separate `num_checkpoints` variable. The `Metric` is only concerned about the metric specification, not the specific interval.

**Review by Adamits — COMMENTED** [review]
Left a few notes, but in general this looks great!

File: yoyodyne/train.py [hunk 5]
```diff
@@ -7,7 +7,7 @@
 import wandb
 from pytorch_lightning import callbacks, loggers
 
-from . import data, defaults, models, schedulers, util
+from . import data, defaults, models, metrics, schedulers, util
```
[the remark this answers is not in the document]
Comment (kylebgorman) [reply yoyodyne/train.py:10]: Good call, fixed.

File: yoyodyne/train.py (hunk 1 above)
Comment (kylebgorman) [reply yoyodyne/train.py:52]: fixed

File: yoyodyne/schedulers.py [hunk 6]
```diff
@@ -133,22 +133,22 @@ def __init__(
...
                 improvement before reducing LR.
             min_learning_rate (float): lower bound on the learning rate.
             **kwargs: ignored.
         """
+        self.metric = metrics.ValidationMetric(reduceonplateau_metric)
         super().__init__(
             optimizer,
-            mode="min" if reduceonplateau_mode == "loss" else "max",
             factor=reduceonplateau_factor,
-            patience=reduceonplateau_patience,
```
[the remark this answers is not in the document]
Comment (kylebgorman) [reply yoyodyne/schedulers.py:150]: That looks to me like a bug so I have reinstated it.

File: yoyodyne/train.py (hunk 4 above)
Comment (kylebgorman) [reply yoyodyne/train.py:53]: +1 yeah

**Comment by kylebgorman** [comment]
Just pushed fixes to everything: you want to take a second look before I merge? (Give me the Approve radio button under the review thingy so I can merge without override...)

**Review by Adamits — APPROVED** [review]
Gave it a quick scroll through after your updates. LGTM
